backend/cmm_helpers.ml

```diff
@@ -731,28 +677,33 @@ let div_int ?dividend_cannot_be_min_int c1 c2 dbg =
             then sub_int q n dbg
             else q
           in
-          let q = asr_const q p dbg in
-          let q_is_negative =
+          let q = asr_const q s dbg in
+          let sign_bit =
+            (* we can use n instead of q when the divisor is non-negative. This
+               makes the instruction dependency graph shallower. *)
             lsr_const (if divisor >= 0n then n else q) (Nativeint.size - 1) dbg
           in
-          add_int q q_is_negative dbg)
+          add_int q sign_bit dbg)
   | _, _ ->
     make_safe_divmod ?dividend_cannot_be_min_int ~if_divisor_is_negative_one
       Cdivi c1 c2 ~dbg
 
 let mod_int ?dividend_cannot_be_min_int c1 c2 dbg =
   let if_divisor_is_positive_or_negative_one ~dividend ~dbg =
-    sequence dividend (Cconst_int (0, dbg))
+    bind "dividend" dividend (fun _ -> Cconst_int (0, dbg))
   in
   match get_const c1, get_const c2 with
-  | _, Some 0n -> sequence c1 (raise_symbol dbg "caml_exn_Division_by_zero")
+  | _, Some 0n -> divide_by_zero c1 ~dbg
   | _, Some (1n | -1n) ->
     if_divisor_is_positive_or_negative_one ~dividend:c1 ~dbg
   | Some n1, Some n2 -> natint_const_untagged dbg (Nativeint.rem n1 n2)
   | _, Some n ->
     if n = Nativeint.min_int
     then
-      (* similarly to the division by min_int almost always being 0, modulo
+      (* [divisor] must be positive be here since we already handled zero and
+         min_int (the only negative power of 2).
```

Comment:
this comment is probably misplaced, given that this section of code deals with `n = min_int` (and `n` is `[divisor]` unless I'm mistaken) which is usually negative. This should instead be a few lines below.